### DISCUSSION TITLE
Best practice to support multiple conversion webhook

### DISCUSSION THREAD

**Discussion by LittleWat** [kubernetes-sigs/kubebuilder#2889] [Support] [answered]
Hello!

I am developing an operator that has three API versions (`v1alpha1`,`v1alpha2`,`v1alpha3`).

It would be easy(?) to do the conversion in steps - convert v1alpha1 to v1alpha2, and then call the v1alpha2->v1alpha3.

But the following document seems to tell us that we should support v1alpha1=>v1alpha3 and v1alpha1=>v1alpha3 conversions
- [Hubs, spokes, and other wheel metaphors - The Kubebuilder Book](https://book.kubebuilder.io/multiversion-tutorial/conversion-concepts.html)

According to this issue,
- [Using conversion-gen with multi-versioning · Issue #1529 · kubernetes-sigs/kubebuilder](https://github.com/kubernetes-sigs/kubebuilder/issues/1529)

`conversion-gen` might help and `cluster-api-provider-aws` utilizes that.

- https://github.com/kubernetes-sigs/cluster-api-provider-aws/tree/main/api
 
These examples do not use multi-steps conversion.
The reason for this is that it is not efficient...?
It would be great if you could share the best practice to support multiple(more than three) conversion webhook.

Thank you in advance.

**Comment by camilamacedo86** [comment, marked as answer]
Hi @LittleWat, 

The K8s docs try to cover it, see: 
- https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#specify-multiple-versions
- https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md
 
Also, as you saw we try to provide an example in the tutorial with: https://book.kubebuilder.io/multiversion-tutorial/tutorial.html

> It would be easy(?) to do the conversion in steps - convert v1alpha1 to v1alpha2, and then call the v1alpha2->v1alpha3.
> But the following document seems to tell us that we should support v1alpha1=>v1alpha3 and v1alpha1=>v1alpha3 conversions

If you have a project version that wants to ensure the desired state on the cluster and that means all be converted to the latest/supported version v1alpha3 why would be better to create many steps to achieve the goal, why not the latest conversion should not doe something like if I find A, B, or C that still support for conversion I will make that for D which is my desired state? 

> It would be great if you could share the best practice to support multiple(more than three) conversion webhook.

I think for example that might be too much right? Also, at some point, you will deprecate and no longer support the old versions.

**Reply by LittleWat** [reply]
Thank you for your reply! I'm sorry for the late reply.

> why would be better to create many steps to achieve the goal

It is because it is not efficient but just easy to implement. For example, when I would like to add the v1alpha3 version, we can reuse the old v1alpha1 to v1alpha2 conversion code and all I need to do is just write the conversion from v1alpha2 to v1alpha3.

> I think for example that might be too much right? Also, at some point, you will deprecate and no longer support the old versions.

Yes, it would be good to stop supporting old versions.

Thank you for your information!

**Comment by camilamacedo86** [comment]
Closing since it is answered.
